### PR TITLE
Allow easy delegates in adapter methods.

### LINKED ISSUE #271: Allow delegates for non-qualified types into adapter methods.
```java
final class PointJsonAdapter {
  @FromJson Point fromJson(JsonReader reader, JsonAdapter<Point> delegate) throws Exception {
    reader.beginObject();
    reader.nextName();
    Point point = delegate.fromJson(reader);
    reader.endObject();
    return point;
  }
}
```
For the above to work, I would need a `@JsonQualifier` on all of my application-layer `Point`s, or I would need to write the `JsonAdapter.Factory` manually.

Should Moshi support the above?

### PR DESCRIPTION
**Pull request by NightlyNexus**
Closes #271.

Are there any more test cases I might not be thinking of?

### CODE REVIEW

**Comment by NightlyNexus** [comment]
tangential: Should we update the IllegalArgumentException messages in AdapterMethodsFactory to state that JsonAdapters can be used?

File: moshi/src/test/java/com/squareup/moshi/AdapterMethodsTest.java [hunk 1]
```diff
@@ -304,15 +343,15 @@
...
+    @FromJson int jsonToInt(JsonReader reader) throws IOException {
       if (reader.peek() == JsonReader.Token.NULL) {
         reader.nextNull();
         return -1;
       }
       return reader.nextInt();
     }
 
-    @ToJson void jsonToInt(JsonWriter writer, int value) throws IOException {
+    @ToJson void intToJson(JsonWriter writer, int value) throws IOException {
```
Comment (NightlyNexus) [inline moshi/src/test/java/com/squareup/moshi/AdapterMethodsTest.java]: sneaking these name fixes in, if that's alright.

**Comment by NightlyNexus** [comment]
You've always got to have the JsonReader and JsonWriter, so you can't do something like:
```
final class PointJsonAdapter {
  @FromJson Point fromJson(JsonAdapter<Point> delegate, String json) throws IOException {
    return delegate.fromJson(json);
  }
}
```
That's a good thing, right?

**Comment by NightlyNexus** [comment]
>tangential: Should we update the IllegalArgumentException messages in AdapterMethodsFactory to state that JsonAdapters can be used?

Closed with #273.

File: moshi/src/main/java/com/squareup/moshi/AdapterMethodsFactory.java [hunk 2]
```diff
@@ -285,14 +285,18 @@ public AdapterMethod(Type type, Set<? extends Annotation> annotations, Object ad
...
       if (jsonAdapters.length > 0) {
         Type[] parameterTypes = method.getGenericParameterTypes();
         Annotation[][] parameterAnnotations = method.getParameterAnnotations();
         for (int i = adaptersOffset, size = parameterTypes.length; i < size; i++) {
-          jsonAdapters[i - adaptersOffset] = moshi.adapter(
-              ((ParameterizedType) parameterTypes[i]).getActualTypeArguments()[0],
-              jsonAnnotations(parameterAnnotations[i]));
+          Type type = ((ParameterizedType) parameterTypes[i]).getActualTypeArguments()[0];
+          Set<? extends Annotation> jsonAnnotations = jsonAnnotations(parameterAnnotations[i]);
+          if (this.type == type && annotations.equals(jsonAnnotations)) {
```
Comment (swankjesse) [inline moshi/src/main/java/com/squareup/moshi/AdapterMethodsFactory.java]: this should probably be Types.equal(). In general neither `==` nor `.equals()` works on `java.lang.reflect.Type`. (for example, `String[].class` isn’t equal to the parameter of `List<String[]>.class`)

File: moshi/src/main/java/com/squareup/moshi/AdapterMethodsFactory.java [hunk 3]
```diff
@@ -285,14 +285,18 @@ public AdapterMethod(Type type, Set<? extends Annotation> annotations, Object ad
...
+            jsonAdapters[i - adaptersOffset] = moshi.nextAdapter(factory, type, jsonAnnotations);
```
Comment (swankjesse) [inline moshi/src/main/java/com/squareup/moshi/AdapterMethodsFactory.java]: would this work better as a ternary assignment?

File: moshi/src/test/java/com/squareup/moshi/AdapterMethodsTest.java [hunk 4]
```diff
@@ -79,6 +79,45 @@
...
+
+  @Test public void toAndFromWithDelegate() throws Exception {
+    Moshi moshi = new Moshi.Builder().add(new PointJsonAdapterWithDelegate()).build();
+    JsonAdapter<Point> adapter = moshi.adapter(Point.class);
+    Point point = new Point(5, 8);
+    assertThat(adapter.toJson(point)).isEqualTo("{\"x\":5,\"y\":8}");
+    assertThat(adapter.fromJson("{\"x\":5,\"y\":8}")).isEqualTo(point);
+  }
+
+  @Test public void toAndFromWithDelegateWithQualifier() throws Exception {
```
Comment (swankjesse) [inline moshi/src/test/java/com/squareup/moshi/AdapterMethodsTest.java]: would these tests pass if the behavior is wrong?

File: moshi/src/test/java/com/squareup/moshi/AdapterMethodsTest.java (hunk 1 above)
Comment (swankjesse) [inline moshi/src/test/java/com/squareup/moshi/AdapterMethodsTest.java]: Yes please.

File: moshi/src/test/java/com/squareup/moshi/AdapterMethodsTest.java (hunk 4 above)
Comment (swankjesse) [inline moshi/src/test/java/com/squareup/moshi/AdapterMethodsTest.java]: Maybe just customize the behavior somehow? Maybe wrap one of ’em in an array?

File: moshi/src/main/java/com/squareup/moshi/AdapterMethodsFactory.java (hunk 2 above)
Comment (NightlyNexus) [inline moshi/src/main/java/com/squareup/moshi/AdapterMethodsFactory.java]: Oops, yep.
Is there a good way to add a test for this?

File: moshi/src/main/java/com/squareup/moshi/AdapterMethodsFactory.java (hunk 3 above)
Comment (NightlyNexus) [inline moshi/src/main/java/com/squareup/moshi/AdapterMethodsFactory.java]: Yep!

File: moshi/src/test/java/com/squareup/moshi/AdapterMethodsTest.java (hunk 4 above)
Comment (NightlyNexus) [inline moshi/src/test/java/com/squareup/moshi/AdapterMethodsTest.java]: 👍

File: moshi/src/main/java/com/squareup/moshi/AdapterMethodsFactory.java (hunk 2 above)
Comment (swankjesse) [inline moshi/src/main/java/com/squareup/moshi/AdapterMethodsFactory.java]: Creating a JsonAdapter for `String[]` should do it. The reflection API is hazardous.

**Comment by swankjesse** [comment]
Looks good, I’m merging. The `String[]` test is no big deal; the code is right.

File: moshi/src/main/java/com/squareup/moshi/AdapterMethodsFactory.java (hunk 2 above)
Comment (NightlyNexus) [inline moshi/src/main/java/com/squareup/moshi/AdapterMethodsFactory.java]: 👍 #274.

**Pull request merged**